[docs] Migrate Themes overview (#982)
moodle/devdocs: docs/apis/plugintypes/theme/images.md

---
title: Images and icons in themes
tags:
  - Plugins
  - Theme
  - Images
  - Icons
sidebar_position: 1
sidebar_label: Images and icons
---

> One of the theme features is the ability to override any of the standard images within Moodle from within your theme. At this point, let's explore how to utilize your own images within your theme and how to override the images being used by Moodle.

Reviewer: ```suggestion
One of the theme features is the ability to override any of the standard images within Moodle when your theme is in use. At this point, let's explore how to utilize your own images within your theme and how to override the images being used by Moodle.
```

So first up a bit about images within Moodle:

> 1. Images you want to use within your theme **need** to be located within your theme's `pix` directory.

Reviewer: Recommend the term _must_ as it fits with the standard RFC language.

```suggestion
1. Images you want to use within your theme **must** to be located within your theme's `pix` directory.
```

1. You can use sub directories within the `pix` directory of your theme.
1. Images used by Moodle's core are located within the `pix` directory of Moodle.

> 1. Modules, blocks and other plugins should also store their images within a pix directory.

Reviewer: ```suggestion
All plugins should store their images within their own `pix` directory.
```

> So making use of your own images first up. Lets assume you have added two image files to the `pix` directory of your theme.

Reviewer: ```suggestion
The following section assumes that there are two image files in the `pix` directory of a theme named `yourthemename`:
```

- `/theme/yourthemename/pix/imageone.svg`
- `/theme/yourthemename/pix/subdir/imagetwo.png`

> Notice that one image is a SVG image, and the second is a PNG. Also the second image is in a subdirectory.

Reviewer: ```suggestion
The first image is an SVG, and the second a PNG located in a subdirectory.
```

> The following code snippet illustrates how to make use of your images within your layout file so they can be inserted by your layout template.

Reviewer: ```suggestion
The following example illustrates how to make use of these images within your layout file so they can be inserted in your layout template.
```

```php title="theme/yourtheme/layout/somelayout.php"
$templatecontext = [
...

> $imageone => $OUTPUT->pix_url('imageone', 'theme'),
> $imagetwo => $OUTPUT->pix_url('subdir/imagetwo', 'theme'),

Reviewer: Indentation wrong?

```suggestion
    $imageone => $OUTPUT->pix_url('imageone', 'theme'),
    $imagetwo => $OUTPUT->pix_url('subdir/imagetwo', 'theme'),
```

...
];

echo $OUTPUT->render_from_template('theme_yourtheme/somelayout', $templatecontext);
```

A method of Moodle's output library is utilized to generate the URL to the image. It's not too important how that function works, but it is important that it's used, as it's what allows images within Moodle to be overridden.

> **DO NOT** include the image file extension. Moodle will work it out automatically and it will not work if you do include it.

Reviewer: Suggest making this an admonition instead?

```handlebars title="theme/yourtheme/templates/somelayout.mustache"
<img src="{{{imageone}}}" alt="Please give your image alt text or set the role to presentation" width="50" height="50">
<img src="{{{imagetwo}}}" alt="Please give your image alt text or set the role to presentation" width="50" height="50">
```

> The following is how you would use the images from within CSS, SCSS or LESS as background images.

Reviewer: Suggest putting this in a subheading?

Reviewer: Also we can probably ditch less.

```css
.divone {
    background-image: url([[pix:theme|imageone]]);
}

.divtwo {
    background-image: url([[pix:theme|subdir/imagetwo]]);
}
```

> In this case, special notations that Moodle looks for must be used. Whenever Moodle hands out a CSS file it first searches for all `something` tags and replaces them with what is required.

Reviewer: ```suggestion
A placeholder is used within the CSS  to allow use of a pix icon. During the CSS/SCSS compilation, these placeholders are converted into a URL which the browser can fetch and serve.
```

> The final thing to notice with both of the cases above is that at no point is the image file extension included.

Reviewer: Admonitino here.

The reason for this leads us into the next topic, how to override images.

> From within a theme you can VERY easily override any standard image within Moodle by simply adding the replacement image to the theme's pix directory in the same sub directory structure as it is in Moodle.

Reviewer: I'd recommend _emphasis_ or **bold** rather than SHOUTING.

So, for instance, if there is a need to override the following images:

1. `/pix/moodlelogo.png`
1. `/pix/i/user.svg`
1. `/mod/chat/pix/monologo.svg`

Simply add the replacement images to the theme in the following locations:

1. `/theme/themename/pix_core/moodlelogo.png`
1. `/theme/themename/pix_core/i/user.svg`
1. `/theme/themename/pix_plugins/mod/chat/monologo.svg`

:::note

> A `pix_core` directory has been created in the theme to store the replacement images. For a specific activity module like chat, the directory `pix_plugins/mod/chat` is needed. This directory is `pix_plugins` and then the plugin type (mod) and then the plugin name (chat).

Reviewer: Bizarre how templates use the frankenstyle, but images us the path /o\

:::

> Now the other very cool thing to mention is that Moodle looks for not just replacements of the same image type (svg, png, jpg, gif, etc...) but also replacements in any image format. That's why when working with our images above, the image file extension was never specified.

Reviewer: We've mostly been steering clear of overly 'friendly' workding like 'very cool'.

> Now the other very cool thing to mention is that Moodle looks for not just replacements of the same image type (svg, png, jpg, gif, etc...) but also replacements in any image format. That's why when working with our images above, the image file extension was never specified.

Reviewer: etc. => and so on.
